verl/utils/debug/profile.py

```diff
@@ -91,3 +94,99 @@ def stop_trace(self):
         if self.check():
             print(f"[Profiler] Trace stopped for rank {self.rank}")
             self.skip_prof = True
+
+
+@dataclass
+class ProfilerConfig:
+    """
+    Worker profiler config. Currently only support Nsight system profiler.
+    """
+
+    all_ranks: bool = False
+    ranks: Optional[list[int]] = None
+    discrete: bool = False
+
+    def union(self, other: "ProfilerConfig") -> "ProfilerConfig":
+        return ProfilerConfig(
+            all_ranks=self.all_ranks or other.all_ranks,
+            ranks=list(set(self.ranks or []) | set(other.ranks or [])),
+            discrete=self.discrete or other.discrete,
+        )
+
+    def intersect(self, other: "ProfilerConfig") -> "ProfilerConfig":
+        return ProfilerConfig(
+            all_ranks=self.all_ranks and other.all_ranks,
+            ranks=list(set(self.ranks or []) & set(other.ranks or [])),
+            discrete=self.discrete and other.discrete,
+        )
+
+
+class NsightSystemsProfiler:
+    """
+    Nsight system profiler. Installed in a worker to control the Nsight system profiler.
+    """
+
+    def __init__(self, rank: int, config: Optional[ProfilerConfig] = None):
+        config = config or ProfilerConfig()
+        self.this_step: bool = False
+        self.discrete: bool = config.discrete
+        self.this_rank: bool = rank in (config.ranks or []) or config.all_ranks
```

Comment:
For nsight the profiler object should have the counter itself. What about adding the same arguments from torch profiler (warmup, wait, active). Since the state is kept in the profiler, you can remove the start/stop calls from the trainer

Reply:
this way the annotated functions can always call start/stop before and after the function call based on the step count maintained by itself. and you can remove the changes in trainer

Reply:
I would like to support two modes, whole step mode (discrete=false) and task discrete mode. 
In whole step mode, we need a scheme for the controller to inform the workers the training step start and stop.
In discrete mode, the annotated task function can maintain its own step counter, but there are chances that the mulitple counters not are synchronized and increase complexity. For example, we saw in DAPO, the generation may run more times that actor_update due to generation fails to collect enough qualified sequences. It is better to get a synchronized singal of training step from the verl framework as an infrastructure.

Reply:
you can track & config them separately with `Map["function_name"-> step_count]`